src/com/google/javascript/jscomp/js/es6/array/fill.js

```diff
@@ -31,14 +33,19 @@ $jscomp.polyfill('Array.prototype.fill', function(orig) {
    * @suppress {reportUnknownTypes, strictPrimitiveOperators}
    */
   var polyfill = function(value, opt_start, opt_end) {
-    var length = this.length || 0;
-    if (opt_start < 0) {
-      opt_start = Math.max(0, length + /** @type {number} */ (opt_start));
+    'use strict';
+    var length = $jscomp.toLength(this.length);
+    var start = $jscomp.toInteger(opt_start);
+    if (start < 0) {
+      start = Math.max(0, length + start);
     }
-    if (opt_end == null || opt_end > length) opt_end = length;
-    opt_end = Number(opt_end);
-    if (opt_end < 0) opt_end = Math.max(0, length + opt_end);
-    for (var i = Number(opt_start || 0); i < opt_end; i++) {
+    var end = opt_end === undefined ? length : $jscomp.toInteger(opt_end);
```

Comment:
There's a lot of repetitive code here and in other files.  What do you think about any of the following:

* $jscomp.toInteger takes a second optional argument for what to return on undefined (may not be quite as smooth as just checking isNaN tho since we still want NaN to map to zero even if undefined maps to length - so not sure if it ends up being worth it).
* add a $jscomp.toPosition(index, length) that handles negative indices and clamping all in one place (but would be nice to handle undefined here as well, and sometimes that's 0 and sometimes it's length - but at least we could just call $jscomp.toPosition(opt_start !== undefined ? opt_start : length, length))